app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt

```diff
@@ -0,0 +1,124 @@
+        Intents.intended(
+            Matchers.allOf(
+                IntentMatchers.hasComponent(MainActivity::class.java.name),
+                IntentMatchers.toPackage("com.github.multimatum_team.multimatum")
+            )
+        )
+        Intents.release()
+    }
+
+    @Test
+    fun qRDisplayTest() {
+        val value = "Ceci est un test"
+        Espresso.onView(ViewMatchers.withId(R.id.QRTextEdit))
+            .perform(ViewActions.replaceText(value))
+        Espresso.closeSoftKeyboard()
+        Espresso.onView(ViewMatchers.withId(R.id.QRButton)).perform(ViewActions.click())
+
+        Espresso.onView(ViewMatchers.withId(R.id.QRGenerated)).check(matches(withQRCode(value)))
+    }
+    /*
+    Test to check the toast. Only work with robolectric
+
+    @Test
+    fun toastDisplayNoData() {
+        Espresso.onView(ViewMatchers.withId(R.id.QRTextEdit)).perform(ViewActions.replaceText(""))
+        Espresso.closeSoftKeyboard()
+        Espresso.onView(ViewMatchers.withId(R.id.QRButton)).perform(ViewActions.click())
+        assertThat(ShadowToast.getTextOfLatestToast(), equalTo("Enter some data"))
+    }
+    */
+
+    /*
+    QR Code matcher found in:
+    https://github.com/dedis/popstellar/blob/15d2b0f062537774113ef9bdf1f73f92c29239db/fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/detail/LaoDetailActivityTest.java
+
+     */
+    private fun withQRCode(expectedContent: String): Matcher<in View?> {
+        return object : BoundedMatcher<View?, ImageView>(ImageView::class.java) {
+            override fun matchesSafely(item: ImageView): Boolean {
+                val actualContent = extractContent(item)
+                return expectedContent == actualContent
+            }
+
+            override fun describeTo(description: Description) {
+                description.appendText("QRCode('$expectedContent')")
+            }
+
+            override fun describeMismatch(item: Any, description: Description) {
+                if (super.matches(item)) {
+                    // The type is a match, so the mismatch came from the QRCode content
+                    val content = extractContent(item as ImageView)
+                    description.appendText("QRCode('$content')")
+                } else {
+                    // The mismatch is on the type, let BoundedMatcher handle it
+                    super.describeMismatch(item, description)
+                }
+            }
+
+            private fun extractContent(item: ImageView): String {
+                val drawable: Drawable = item.drawable
+                require(drawable is BitmapDrawable) { "The provided ImageView does not contain a bitmap" }
+                val binary = convertToBinary(drawable.bitmap)
+                return try {
+                    // Parse the bitmap and check it against expected value
+                    QRCodeReader().decode(binary).text
+                } catch (e: NotFoundException) {
```

Comment:
Slight nitpick:
It's sad that you can't catch multiple exception types at once but I think you can deduplicate the error message by doing:
```kotlin
try {
    ...
} catch (ex: Exception) {
    when(ex) {
        is NotFoundException, is ChecksumException, is FormatException -> {
            throw IllegalArgumentException("The provided image is not a valid QRCode", e)
        }
        else -> throw ex
    }
}
```
(see <https://stackoverflow.com/a/36760590/5559663>)